domain/keyupdater/state/controllerkeystate.go

```diff
@@ -0,0 +1,73 @@
+package state
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/domain"
+)
+
+// ControllerKeyState provides a state access layer for accessing a controllers
+// ssh keys via controller config.
+type ControllerKeyState struct {
+	*domain.StateBase
+}
+
+// GetControllerConfigKeys returns the controller config key and values for the
+// keys supplied. If one or more keys supplied do not exist in the controller's
+// config they will be omitted from the final result.
+func (st *ControllerKeyState) GetControllerConfigKeys(
```

Comment:
I've seen us implement this exact same method in multiple places. Why don't we have a common state package which defines a `ControllerKeyGetter` type with this method on it, and then other state types that need to access controller config can embed `ControllerKeyGetter`?

Reply:
We made the decision a while back to repeat ourselves over taking a dry approach. This way the state layer can adapt over time to the needs of the service instead of ending up with a func that takes a bunch of tuning knobs.

Reply:
The only use of this is to pass in "ControllerSSHKeys". To avoid proliferation of generic methods that are more heavy weight than needed, we can just implement a bespoke method to precisely extract just the value needed to fit the use case. Until there's a need to query another controller config value, which I don't think we expect, we should err on the side of precise, bespoke sql for each use case.

Reply:
Ordinarily I might agree with you but I think in this case that is the wrong move. If this key was explicitly modelled within the ddl and had meaning in the database I would be all for it. But model config is model as a set of key value attributes that the database has no meaning associated with. It is only the service layer that applies meaning to these values and intent.

Leaving this at the service layer keeps it at the correct level like the rest of the code. By pulling this into state we import another package into the state layer that it doesn't need to be polluted with and we also start giving the state layer business logic that it doesn't have in it's own scoped domain.

Reply:
And in reality the query doesn't change at all. It is just the return value. So very small change to the state layer but big change to our separation of concerns.

Reply:
I can see that argument as well. But then it begs the question - we have modelled user ssh keys correctly in a user_public_ssh_key table. Why don't we model the controller ssh key correctly as well. We've pulled out secret backend from model config, and will be doing so for default space. Now would be the time to bite the bullet and try and slim down the various config data bags.

Reply:
Cheers and I agree on this. I just can't go off on a side quest for this. I got approval to model authorised keys this way to get it done instead of using model config. Only so much scope I can be afforded for refactor.

Reply:
So long as we add a jira card / update risk register and commit to following up on this sooner rather than later, then we can make progress with this PR. Can we at least add a todo with an explanation.